docs(editor): add DataBinding Lists (#139)
rive-app/rive-docs: editor/data-binding/lists.mdx

---
title: 'Lists'

> description: 'Use DataBinding to generate lists at runtime'

Reviewer: ```suggestion
description: 'Use Data Binding to generate lists at runtime'
```

---

> <Note>Lists are currently available in <a href="https://rive.app/blog/early-access-to-unreleased-features" target="_blank" rel="noopener noreferrer">Early Access</a> as a feature in the Rive Editor. Further performance enhancements as well as Runtime APIs will be available when released to the general public.</Note>

Author: Thanks for the suggestion, done!

A List is a way to display a set of items that are dynamically generated based on bound data values that you set up in your View Models. This allows you to build Rive files that can change in real time based on updates to those values. You can use Lists to create:

- Menus with a dynamic amount of options
- Product listings
- Notifications or activity feeds
- Chat messages
- Dropdown menus
- Contacts, friends, or followers lists
- High scores, tables, and more

---

## Artboard Lists

Artboard Lists enable you to generate a number of list items using an Artboard to represent each item in your List. Artboard Lists must be added as children of an Artboard or Layout. 

To add an Artboard List to the stage, first select either an Artboard or a Layout. In the inspector on the right side of the editor, you will see a button to add an Artboard List. Click it to add the List to your hierarchy. It will appear as a child of the Artboard or Layout had previously selected.

![Image](/images/editor/add-artboard-list.png)

Once the Artboard List is added to your hierarchy, you can select it and see its inspector.

![Image](/images/editor/artboard-list-property.png)

> The next step is to bind data to it using DataBinding. This will determine the content and number of items in your List. There are 2 ways to generate List content:

Reviewer: ```suggestion
The next step is to bind data to it using Data Binding. This will determine the content and number of items in your List. There are 2 ways to generate List content:
```

- Using the View Model List property
- Using a View Model Number property together with a Number to List converter

## View Model List Property

> <Note>Before continuing, it's important to understand the fundamentals of DataBinding, in particular, what a View Model is, how to create one and how to bind it to an object's properties. Learn more in the [DataBinding Overview](/editor/data-binding/overview).</Note>

Reviewer: ```suggestion
<Note>Before continuing, it's important to understand the fundamentals of Data Binding, in particular, what a View Model is, how to create one and how to bind it to an object's properties. Learn more in the [Data Binding Overview](/editor/data-binding/overview).</Note>
```

> A View Model List is a property that can contain N items which each represent a View Model instance. In order to be used in a List, the View Model must be bound to an Artboard.

Reviewer: ```suggestion
A View Model List is a property that can contain a dynamic number of items which each represent a View Model instance. In order to be used in a List, the View Model must be bound to an Artboard.
```

I'm unclear whether the concept of "N number" is an engineering or design or otherwise concept, so I only suggest this for clarification.

3. Bind that View Model to an Artboard that you want to use to render your List item. This is where you may also want to create additional View Model properties and bind those to object properties on that Artboard.
4. Click the + button beside View Models to create another View Model (this will be the View Model that contains your List and should be bound to the Artboard where you want your List to render)
5. Select the newly created View Model and click the + button next to it
6. From the popout select **List**. This adds a List property to the View Model
![Image](/images/editor/list-property.png)
7. Select the List property and in the inspector on the right, you can add items by clicking Add List item
![Image](/images/editor/list-items.png)
8. Once a List item is added, you can click the settings button to the left of its name in order to set the View Model and View Model instance for that item
9. After adding your List items, go back to the Hierarchy tab, select the Artboard List, and from the Artboard List property dropdown, select the ViewModel List property you created above. 

> 10. Run the state machine and you should see your list items. Remember that the layout will be determined by the Artboard List's parent, so modify the parent's settings in order to tweak your List's layout

Reviewer: Just as a callout, there is a [Steps](https://mintlify.com/docs/components/steps#steps) component that can be used in Mintlify!

Author: Nice, using steps looks MUCH better! Done!

> <Note>Rive's runtimes will provide APIs to modify the List and List items at runtime (for example, adding or removing items), however this is not yet available in Early Access.</Note>

Reviewer: ```suggestion
<Warning>Rive's runtimes will provide APIs to modify the List and List items at runtime (for example, adding or removing items), however this is not yet available in Early Access.</Warning>
```

## View Model Number with Converter

> The second way to populate your list is by specifying the number of items you want in your list along with the ViewModel (Artboard) you want to instance. This can be done using a View Model Number property in combination with the new Number to List converter.

Reviewer: ```suggestion
The second way to populate your list is by specifying the number of items you want in your List along with the ViewModel (Artboard) you want to instance. This can be done using a View Model Number property in combination with the new Number to List converter.
```

> To Create a View Model Number and converter and bind it to an Artboard List:

Reviewer: ```suggestion
To Create a View Model Number to List converter and bind it to an Artboard List:
```
